crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/meta_net.rs

```diff
@@ -824,23 +861,67 @@ impl MetaNet {
             preflight: Some((
                 Arc::new(move |_| {
                     let i_s = kitsune_internal_sender.clone();
+                    let user_data_bytes_sent = user_data_bytes_sent_clone.clone();
 
                     Box::pin(async move {
                         let agent_list = i_s
                             .get_all_local_joined_agent_infos()
                             .await
                             .unwrap_or_default();
-                        wire::Wire::peer_unsolicited(agent_list).encode_vec()
+                        PreflightData::v0(
+                            KITSUNE_PROTOCOL_VERSION,
+                            agent_list,
+                            user_data_bytes_sent,
+                        )
+                        .encode_vec()
                     })
                 }),
-                Arc::new(move |_, data| {
+                Arc::new(move |url, data| {
                     let e_s = evt_sender.clone();
-                    Box::pin(async move {
-                        match wire::Wire::decode_ref(&data) {
-                            Ok((
-                                _,
-                                wire::Wire::PeerUnsolicited(wire::PeerUnsolicited { peer_list }),
-                            )) => {
+                    let url = url.clone();
+                    let user_data_sent = user_data_sent.clone();
+                    let user_data_bytes_sent = user_data_bytes_sent.clone();
+                    match PreflightData::decode_ref(&data) {
+                        Ok((
+                            _,
+                            PreflightData::V0(V0 {
+                                kitsune_protocol_version,
+                                peer_list,
+                                user_data: user_data_bytes_received,
+                            }),
+                        )) => {
+                            Box::pin(async move {
+                                if kitsune_protocol_version != KITSUNE_PROTOCOL_VERSION {
+                                    tracing::warn!(
+                                        ?url,
+                                        "kitsune protocol version mismatch: ours = {}, theirs = {}",
+                                        KITSUNE_PROTOCOL_VERSION,
+                                        kitsune_protocol_version,
+                                    );
+                                    return Ok(());
+                                }
+
+                                if user_data_bytes_received != user_data_bytes_sent {
```

Comment:
Would something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=58f0e3d766993822feeb86f3db799155) make it respect the None condition matching, while at the same time eliminate the need for that Generic param?